text/0000-argumentless-fn-helper-returns-noop-function.md

```diff
@@ -0,0 +1,278 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- Relevant Team(s): (fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies)
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Return noop function if {{fn}} helper is invoked without arguments
+
+## Summary
+
+This RFC introduces an argumentless version of the `{{fn}}` template helper.
+It should return a noop function.
+
+The RFC does not intend to change the existing meaning of the `{{fn}}` template
+helper if invoked with one or more arguments. Especially it does *not* propose
+to support `undefined` as value for the first positional argument. The `{{fn}}`
+template helper should continue to throw an assertion if invoked with another
+value than a function as the first positional argument.
+
+## Motivation
+
+This RFC is driven by two motivations:
+
+1. Provide a built-in way to get a noop function in a template.
+2. Simplify mental model around `{{fn}}` helper.
+
+### Use cases for noop functions in templates
+
+Noop function are commonly used in modern Ember. Currenlty the framework misses
+a built-in way to get one in a template. Developers need to either reach out
+for a userland tempalte helper like the [`{{noop}}` helper from
+ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers#noop)
+package or expose a noop function from backing JavaScript class.
+
+A noop function is mainly required in components that have an optional
+argument, which should be either passed down to another component as an action
+handler or used with an event listener registered through `{{on}}` modifier.
+The value of the optional argument is `undefined` if consumer does not set it
+but the component may and the `{{on}}` modifier do not support `undefined` but
+require a function.
+
+Let's have a look at two example use cases to illustrate the issue:
```

Comment:
Another place I frequently ready for `noop` is in integration tests when invoking a component with a passed action that isn't needed in the context of the specific test:
```
await render(hbs`<VocabulariesExpanded
  @vocabularies={{this.vocabularies}}
  @collapse={{noop}}
/>`);
```

Reply:
I get what you mean. But I also experienced that there is value in being more explicit about if the function should be called or if not. You may miss some bugs if you don't assert that a function is *not* executed if it should not be executed.

Reply:
Makes sense. I tent to treat that as a unique unit in an isolated test. Either way, I like empty `fn` a lot.